README.md
python-manta is a community-maintained Python SDK for [the Joyent Manta
Object Storage Service](http://www.joyent.com/manta) (a.k.a. Manta). This
provides a Python 'manta' package (for using the [Manta REST
API](http://apidocs.joyent.com/manta/api.html) and a 'mantash' (MANTA SHell) CLI
and shell. For an introduction to Manta in general, see [Manta getting started
docs](http://apidocs.joyent.com/manta/index.html).


# Current Status

Tested mostly on Mac and SmartOS using Python 2.7. Linux should work.
The *intention* is to support Windows as well. Python 3 support is new and we
welcome feedback.

Python 2.6 support has been dropped in python-manta 2.7.

Feedback and issues here please: <https://github.com/joyent/python-manta/issues>


# Installation

tl;dr: `pip install --upgrade manta`


## 0. install `pip` (and maybe pynacl)

**SmartOS**:

1. Install pip:
    pkgin install py27-pip
    or
    pkgin install py34-pip

2. Install libsodium and pynacl
    pkgin install libsodium
    SODIUM_INSTALL=system pip install pynacl

**Mac**:

    # See <http://www.pip-installer.org/en/latest/installing.html>
    curl -O https://bootstrap.pypa.io/get-pip.py
    sudo python get-pip.py

**Ubuntu**:

    sudo apt-get install python3-pip 
    or
    sudo apt-get install python-pip

Others? Please [let me know](https://github.com/joyent/python-manta/issues/new)
if there are better instructions that I can provide for your system, so I can
add them here.



## 1. install python-manta

The preferred way is:

    pip install manta

If you don't have pip (see above), but have `easy_install` then:

    easy_install manta

You should also be able to install from source:

    git clone https://github.com/joyent/python-manta.git
    cd python-manta
    python setup.py install    # might require a 'sudo' prefix


## 2. verify install worked

The 'mantash' CLI should now work:

    $ mantash help
    ...

And `import manta` should now work:

    $ python -c "import manta; print(manta.__version__)"
    2.0.0



# Setup

First setup your environment to match your Joyent Manta account. Adjust
accordingly for your SSH key and Manta login. The SSH key here must match
one of keys uploaded for your Joyent Public Cloud account.

    export MANTA_KEY_ID=`ssh-keygen -l -f ~/.ssh/id_rsa.pub | awk '{print $2}'`
    export MANTA_URL=https://us-east.manta.joyent.com
    export MANTA_USER=jill
    export MANTA_SUBUSER=bob # optional, if using RBAC subuser
    export MANTA_ROLE=ops # optional, if specifying a non-default role for the subuser

`mantash` uses these environment variables (as does the [Manta Node.js SDK
CLI](https://apidocs.joyent.com/manta/index.html#setting-up-your-environment)).
Alternatively you can specify these parameters to `mantash` via command-line
options -- see `mantash --help` for details.

For a colourful `mantash` prompt you can also set:

    export MANTASH_PS1='\e[90m[\u@\h \e[34m\w\e[90m]$\e[0m '

or more simply:

    export MANTASH_PS1='[\u@\h \w]$ '

See `_update_prompt` in bin/mantash for the list of supported PS1 escape
codes.


Now test that things are working:

    $ mantash ls /$MANTA_USER
    jobs
    public
    reports
    stor

If not, see for the [Troubleshooting](#troubleshooting) section below.


# Python Usage

```python
import os
import logging
import manta

# Manta logs at the debug level, so the logging env needs to be setup.
logging.basicConfig()

url = os.environ['MANTA_URL']
account = os.environ['MANTA_USER']
key_id = os.environ['MANTA_KEY_ID']

# optional fields for RBAC
subuser = os.environ.get('MANTA_SUBUSER', None)
role = os.environ.get('MANTA_ROLE', None)

# This handles ssh-key signing of requests to Manta. Manta uses
# the HTTP Signature scheme for auth.
# http://tools.ietf.org/html/draft-cavage-http-signatures-00
signer = manta.SSHAgentSigner(key_id)

client = manta.MantaClient(url, account, subuser=subuser,
                           role=role, signer=signer)

content = client.get_object('/%s/stor/foo.txt' % account)
print content

print dir(client)   # list all methods, better documentation coming (TODO)
```

See more examples in the [examples/](./examples/) directory.


# CLI

This package also provides a `mantash` (MANTA SHell) CLI for working with
Manta:

```shell
$ mantash help
Usage:
    mantash COMMAND [ARGS...]
    mantash help [COMMAND]
...
Commands:
    cat            print objects
    cd             change directory
    find           find paths
    get            get a file from manta
    job            Run a Manta job
...

# This is a local file.
$ ls
numbers.txt

# Mantash single commands can be run like:
#       mantash ls
# Or you can enter the mantash interactive shell and run commands from
# there. Let's do that:
$ mantash
[jill@us-east /jill/stor]$ ls
[jill@us-east /jill/stor]$                      # our stor is empty
[jill@us-east /jill/stor]$ put numbers.txt ./   # upload local file
[jill@us-east /jill/stor]$ ls
numbers.txt
[jill@us-east /jill/stor]$ cat numbers.txt
one
two
three
four

# List available commands. A number of the typical Unix-y commands are
# there.
[jill@us-east /jill/stor]$ help
...

# Manta jobs.
#
# Note: The '^' is used as an alternative pipe separator to '|'.
# The primary reason is to avoid Bash eating the pipe when running
# one-off `mantash job ...` commands in Bash.

# Run a Manta job. Here `grep t` is our map phase.
[jill@us-east /jill/stor]$ job numbers.txt ^ grep t
two
three

# Add a reduce phase, indicated by '^^'.
[jill@us-east /jill/stor]$ job numbers.txt ^ grep t ^^ wc -l
2
```



# License

MIT. See [LICENSE](./LICENSE).

Some pure Python dependencies are included in this distribution (to
reduce install dependency headaches). They are covered by their
respective licenses:

- paramiko (http://www.lag.net/paramiko/): LGPL
- httplib2 (http://code.google.com/p/httplib2/): MIT
- cmdln (https://github.com/trentm/cmdln): MIT
- appdirs (https://github.com/ActiveState/appdirs): MIT


# Limitations

The python-manta Python API isn't currently well-suited to huge objects
or huge directory listings (>10k dirents) because responses are fully
buffered in memory rather than being streamed. If streaming is a requirement
for your use case, you could consider the [Manta Node.js
bindings](https://github.com/joyent/node-manta).

For other limitations (also planned work) see TODO.txt.


# Troubleshooting

An attempt to cover some common install/setup issues.

## `pynacl` dependency install failure on SmartOS

This test failure:
```
    PASS: pwhash_argon2id
    /tmp/pip-install-AIWK8Y/pynacl/src/libsodium/build-aux/test-driver: line 107: 53648: Memory fault(coredump)
    FAIL: randombytes
    PASS: scalarmult
```

Ultimately results in this install error:
```

  File "/opt/local/lib/python2.7/subprocess.py", line 186, in check_call
      raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', 'check']' returned non-zero exit status 2

            ----------------------------------------
            Command "/opt/local/bin/python2.7 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-AIWK8Y/pynacl/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-YflZ4O/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-AIWK8Y/pynacl/

```

See: https://github.com/joyent/python-manta/issues/55

## `x509 certificate routines:X509_load_cert_crl_file` error

```
$ mantash ls
mantash: ERROR: [Errno 185090050] _ssl.c:343: error:0B084002:x509 certificate routines:X509_load_cert_crl_file:system lib (/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.py:141 in __init__)

Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/manta-2.4.1-py2.7.egg/EGG-INFO/scripts/mantash", line 2001, in <module>
    retval = main(sys.argv)
...
  File "/Library/Python/2.7/site-packages/httplib2-0.8-py2.7.egg/httplib2/__init__.py", line 80, in _ssl_wrap_socket
    cert_reqs=cert_reqs, ca_certs=ca_certs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.py", line 387, in wrap_socket
    ciphers=ciphers)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.py", line 141, in __init__
    ciphers)
SSLError: [Errno 185090050] _ssl.c:343: error:0B084002:x509 certificate routines:X509_load_cert_crl_file:system lib
```

This is saying that python-manta (the httplib2 package it is using) cannot
verify the MANTA\_URL server certificate. In some cases the problem here is
write access to the "cacerts.txt" file in the installed httplib2 package. That
can be solved by making that file world readable (as [discussed
here](http://stackoverflow.com/a/19145997)).

```
$ sudo chmod 644 $(python -c 'from os.path import dirname; import httplib2; print dirname(httplib2.__file__)')/cacerts.txt
Password:
```

# Development and Testing Notes

In order to make sure testing covers RBAC, you'll want to make sure you have a
subuser set up with appropriate permissions for Manta in addition to the
environment variable setup described above.

```
mkdir ./tmp

# create a dedicated test user
sdc-user create --login=python_manta --password=${PASSWORD} --email=${EMAIL}

# create a new ssh key and upload it for our user
ssh-keygen -t rsa -b 4096 -C "${EMAIL}" -f ./tmp/python_manta
sdc-user upload-key \
         $(ssh-keygen -E md5 -lf ./tmp/manta | awk -F' ' '{gsub("MD5:","");{print $2}}') \
         --name=python_manta python_manta ./tmp/python_manta.pub

# create a policy with minimum permissions we need
sdc-policy create --name=python_manta \
           --rules='CAN putdirectory' \
           --rules='CAN listdirectory' \
		   --rules='CAN getdirectory' \
		   --rules='CAN deletedirectory' \
           --rules='CAN putobject' \
	       --rules='CAN putmetadata' \
           --rules='CAN getobject' \
		   --rules='CAN deleteobject' \
		   --rules='CAN putsnaplink'

# create a new role with that policy and attach it to our user
sdc-role create --name=python_manta \
		--policies=python_manta \
		--members=python_manta

# create a directory with our role assigned to it
mmkdir ${MANTA_USER}/stor/tmp --role-tag=python_manta
```


## Release process

Here is how to cut a release:

1. Make a commit to set the intended version in
   [manta/version.py](manta/version.py) and changing `## not yet released` at
   the top of "CHANGES.md" to:

    ```
    ## not yet released

    (nothing yet)


    ## $version
    ```

   Run `make versioncheck` to ensure you have the version updated correctly.

2. Commit and push that change.

3. Use the following make target to do the release:

    ```
    make cutarelease
    ```

   This will run a couple checks (clean working copy, versioncheck) and
   then will git tag and publish to pypi. If the PyPI upload fails, you can
   retry it via `make pypi-upload`.


CHANGES.md
# python-manta Changelog

## not yet released

(nothing yet)


## 3.0.0

- [issue #58] Add Python 3 support.

## 2.7.0

- [issue #50, pull #56] Add support for ECDSA keys for auth (by Rui Loura).
- [issue #51] Fix an issue with `MANTA_KEY_ID=MD5:...` handling (i.e. an
  MD5 key that uses the now more common "MD5:" prefix.
- [issue #54] Fix a crash in `mantash find DIR -name NAME ...`.


## 2.6.1

- [issue #41] Support `MANTA_KEY_ID=SHA256:...` fingerprint form.
- [issue #43] Ensure a broken `~/.ssh/*.pub` symlink doesn't break
  finding a key for a given `MANTA_KEY_ID`.


## 2.6.0

- [MANTA-2925] `mantash cp [-R] ...` command for copying files within Manta
  (including recursive dir copy).


## 2.5.0

- [pull #35] RBAC support (see: https://docs.joyent.com/public-cloud/rbac)
  The `MantaClient` constructor now takes optional `subuser` and `role`
  parameters. The `mantash` CLI gets the params from either the `--user` /
  `--role` command line arguments or from the `MANTA_SUBUSER` / `MANTA_ROLE`
  environment vars.  This behavior matches the Node.js Manta client.  By Tim
  Gross.

- [pull #29] Fix doc links.


## 2.4.1

- [pull #34] Fix handling of the keyId/`MANTA_KEY_ID` to support being set to a
  non-MD5 fingerprint. By default Linux (and more recently Mac) versions of OpenSSH's
  `ssh-keygen -lf ...` show a SHA256 form of the fingerprint.
- [pull #34] Fix a bug where `mantash mkdir -p /trent.mick/public/a/b/c` wouldn't
  create the top `a` directory (by Tim Gross).
- Fix `mantash du ...` to sort output for a given dir by basename.
- Fix a logic bug in `mantash mkdir -p ...` where the last dir could possibly
  not be created.


## 2.4.0

- `mantash du ...`


## 2.3.0

- `mantash rm PATH ...` now supports globs.
- Allow the second arg to `mantash ln` to be optional a la typical `/bin/ln`.
- Fix usage of `\n` in `MANTASH_PS1` envvar.


## 2.2.0

- [MANTA-2530] Add `MantaClient.{get_job,get_job_input,get_job_output,
  get_job_failures,get_job_errors}` wrappers of the `RawMantaClient` to
  handle retrieving [*archived job
  results*](https://apidocs.joyent.com/manta/jobs-reference.html#job-completion-and-archival).
- Change `mantash jobs` to emit a flat list of job IDs by default
  (like `mantash ls`). Use `-j|--json` to get a JSON list.
- Add `-v` option to all the examples for easier verbose output.
- [issue #23] `cd foo/bar` where no "foo" dir exists crashes mantash
- Add 'vim' alias to 'vi' in mantash.


## 2.1.1

- [pull #14] Fix `fingerprint_from_ssh_pub_key()` parsing of SSH keys
  which caused fingerprint calculation errors for *some* SSH keys with
  comments. (Thanks <https://github.com/tomahn>!)

- [issue #20] Change cache dirs used by `mantash` and the client to include
  the effetive UID. This avoids permission issues if one runs as root,
  e.g. via `sudo -E mantash ...`, once.

- [issue #19] Paramiko 1.14.0 changed `<Agent key>.sign_ssh_data()` signature
  which broke request signing via the ssh-agent. Fix that.

- Add limited `mantash mv` command for moving Manta objects and dirs around.

- Fix test suite breakage.


## 2.1.0

- [issue #18] Add MANTA_NO_AUTH support for running python-manta in a Manta
  job or mlogin session. For example, this means that you can now much more
  easily use python-manta in a manta job by using `pip install manta` in
  your init command. E.g. using the node-manta `mjob` command:

        mfind /bob/stor/datafiles -to \
            | mjob create --init 'pip install manta' \
                -s /bob/stor/scripts/my-processor.py \
                -m '/assets/bob/stor/scripts/my-processor.py'

  Now "my-processor.py" can `import manta` to read and write to Manta inside the
  job.

- [issue #16] A start at some encoding handling fixes in `mantash` for paths
  with non-ascii characters.

- [issue #15] Fix tab completion in `mantash` to escape shell special
  characters.

- 'mantash sign MANTA-PATH' wrapper around msign from the node-manta tools.

- 'mantash json' simple command to JSON-pretty-print a smallish JSON file.

- 'mantash head' and 'mantash tail' first cuts.

- Fix mantash bash completion to append '/' for directories (both for manta dirs
  and local dirs as appropriate for the command).

- Add 'mantash login ...' command. This calls out to the awesome
  [`mlogin`](https://apidocs.joyent.com/manta/mlogin.html) tool from the [Manta
  Node.js SDK (aka node-manta)](https://github.com/joyent/node-manta), hence
  this requires that you have node-manta setup and on your PATH.

- Sort 'mantash find ...' output.

- Add support to 'mantash' to remember user areas visited for future
  tab completion. For example, if you are "bob" and you've visited
  "/sue/public/..." then "cd /s<TAB>" will complete to "cd /sue/" for you.


## 2.0.1

- Packaging tweaks, improved README, clean rev for pypi.


## 2.0.0

- Rework the packaging for `pip install manta` et al to work.
  Drop siloed 3rd party deps. This work inspired and mostly
  copied from [deserat's](https://github.com/deserat) pull #10.


## 1.5.1

- [issue #12] Fixes for `MantaClient.put_object`.


## 1.5.0

- [issue #9, MANTA-1593] Proper URI encoding (RFC3986).


## 1.4.0

- [MANTA-1478] Properly handle paging through ListDirectory results. Before
  this, `<MantaClient>.ls(...)` and `mantash ls` would not return all
  entries in a directory with greater than ~256 entries.

- Add 'mantash open MANTA-PATH' to open a file in Manta in your browser.

- Rename 'mantash gzcat ...' to 'mantash zcat ...' as it should have been.

- [issue #8] A start at Python 3 support. This is **incomplete**. Paramiko
  doesn't support Python 3 so we are stuck. Update to httplib2 0.8.


## 1.3.1

- Fix parsing of DSA pubkeys when generating a pubkey fingerprint
  for `MANTA_KEY_ID=<path to ssh private key>` usage.


## 1.3.0


manta/auth.py
# Copyright 2018 Joyent, Inc.  All rights reserved.

"""Manta client auth."""

from __future__ import absolute_import
import binascii
import sys
import io
import os
from os.path import expanduser
import logging
import base64
import hashlib
from getpass import getpass
import re
import struct
from glob import glob

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.hashes import SHA1, SHA256, SHA384, SHA512
from cryptography.hazmat.primitives.asymmetric import padding, ec, rsa
from cryptography.hazmat.primitives.asymmetric.utils import encode_dss_signature

from paramiko import Agent
from paramiko.message import Message

from manta.errors import MantaError


#---- globals

log = logging.getLogger('manta.auth')

FINGERPRINT_RE = re.compile(r'(^(MD5:)?([a-f0-9]{2}:){15}[a-f0-9]{2})|(^SHA256:[a-zA-Z0-9\/\+]{43})$')

PEM_STRING_RSA_RE = re.compile(r'RSA (PUBLIC|PRIVATE) KEY')
PEM_STRING_ECDSA_RE = re.compile(r'EC(DSA)? (PUBLIC|PRIVATE) KEY')

ECDSA_SHA256_STR = "ecdsa-sha256"
ECDSA_SHA384_STR = "ecdsa-sha384"
ECDSA_SHA512_STR = "ecdsa-sha512"
RSA_STR = "rsa-sha1"

# TODO: add and test other SSH key types
ALGO_FROM_SSH_KEY_TYPE = {
    "ecdsa-sha2-nistp256": ECDSA_SHA256_STR,
    "ecdsa-sha2-nistp384": ECDSA_SHA384_STR,
    "ecdsa-sha2-nistp521": ECDSA_SHA512_STR,
    "ssh-rsa": RSA_STR
}

ECDSA_ALGO_FROM_KEY_SIZE = {
    "256": ECDSA_SHA256_STR,
    "384": ECDSA_SHA384_STR,
    "521": ECDSA_SHA512_STR
}


#---- internal support stuff


def fingerprint_from_ssh_pub_key(data):
    """Calculate the fingerprint of SSH public key data.

    >>> data = "ssh-rsa AAAAB3NzaC1y...4IEAA1Z4wIWCuk8F9Tzw== my key comment"
    >>> fingerprint_from_ssh_pub_key(data)
    '54:c7:4c:93:cf:ff:e3:32:68:bc:89:6e:5e:22:b5:9c'

    Adapted from <http://stackoverflow.com/questions/6682815/>
    and imgapi.js#fingerprintFromSshpubkey.
    """
    data = data.strip()

    # Let's accept either:
    # - just the base64 encoded data part, e.g.
    #   'AAAAB3NzaC1yc2EAAAABIwAA...2l24uq9Lfw=='
    # - the full ssh pub key file content, e.g.:
    #   'ssh-rsa AAAAB3NzaC1yc2EAAAABIwAA...2l24uq9Lfw== my comment'
    if (re.search(r'^ssh-(?:rsa|dss) ', data) or
        re.search(r'^ecdsa-sha2-nistp(?:[0-9]+)', data)):
        data = data.split(None, 2)[1]

    key = base64.b64decode(data)
    fp_plain = hashlib.md5(key).hexdigest()
    return ':'.join(a + b for a, b in zip(fp_plain[::2], fp_plain[1::2]))


def fingerprint_from_raw_ssh_pub_key(key):
    """Encode a raw SSH key (string of bytes, as from
    `str(paramiko.AgentKey)`) to a fingerprint in the typical
    '54:c7:4c:93:cf:ff:e3:32:68:bc:89:6e:5e:22:b5:9c' form.
    """
    fp_plain = hashlib.md5(key).hexdigest()
    return ':'.join(a + b for a, b in zip(fp_plain[::2], fp_plain[1::2]))


def sha256_fingerprint_from_ssh_pub_key(data):
    data = data.strip()

    # accept either base64 encoded data or full pub key file,
    # same as `fingerprint_from_ssh_pub_key`.
    if (re.search(r'^ssh-(?:rsa|dss) ', data) or
        re.search(r'^ecdsa-sha2-nistp(?:[0-9]+)', data)):
        data = data.split(None, 2)[1]

    digest = hashlib.sha256(binascii.a2b_base64(data)).digest()
    encoded = base64.b64encode(digest)  # ssh-keygen strips thi
    return 'SHA256:' + encoded.decode('utf-8')

def sha256_fingerprint_from_raw_ssh_pub_key(raw_key):
    """Encode a raw SSH key (string of bytes, as from
    `str(paramiko.AgentKey)`) to a fingerprint in the SHA256 form:
        SHA256:j2WoSeOWhFy69BQ39fuafFAySp9qCZTSCEyT2vRKcL+s
    """
    digest = hashlib.sha256(raw_key).digest()
    h = base64.b64encode(digest).decode('utf-8')
    h = h.rstrip().rstrip('=')  # drop newline and possible base64 padding
    return 'SHA256:' + h

def load_ssh_key(key_id, skip_priv_key=False):
    """
    Load a local ssh private key (in PEM format). PEM format is the OpenSSH
    default format for private keys.

    See similar code in imgapi.js#loadSSHKey.

    @param key_id {str} An ssh public key fingerprint or ssh private key path.
    @param skip_priv_key {boolean} Optional. Default false. If true, then this
        will skip loading the private key file and `priv_key` will be `None`
        in the retval.
    @returns {dict} with these keys:
        - pub_key_path
        - fingerprint
        - priv_key_path
        - priv_key
        - algorithm
    """
    priv_key = None

    # If `key_id` is already a private key path, then easy.
    if not FINGERPRINT_RE.match(key_id):
        if not skip_priv_key:
            f = io.open(key_id, 'rb')
            try:
                priv_key = f.read()
            finally:
                f.close()
        pub_key_path = key_id + '.pub'
        f = io.open(pub_key_path, 'r')
        try:
            pub_key = f.read()
        finally:
            f.close()
        fingerprint = fingerprint_from_ssh_pub_key(pub_key)

        # XXX: pubkey should NOT be in PEM format.
        try:
            algo = ALGO_FROM_SSH_KEY_TYPE[pub_key.split()[0]]
        except KeyError:
            raise MantaError("Unsupported key type for: {}".format(key_id))

        return dict(
            pub_key_path=pub_key_path,
            fingerprint=fingerprint,
            priv_key_path=key_id,
            priv_key=priv_key,
            algorithm=algo)

    # Else, look at all pub/priv keys in "~/.ssh" for a matching fingerprint.
    fingerprint = key_id

    pub_key_glob = expanduser('~/.ssh/*.pub')
    pub_key = None
    for pub_key_path in glob(pub_key_glob):
        try:
            f = io.open(pub_key_path, 'r')
        except IOError:
            # This can happen if the .pub file is a broken symlink.
            log.debug("could not open '%s', skip it", pub_key_path)
            continue
        try:
            pub_key = f.read()
        finally:
            f.close()

        # The MD5 fingerprint functions return the hexdigest without the hash
        # algorithm prefix ("MD5:"), and the SHA256 functions return the
        # fingerprint with the prefix ("SHA256:").  Ideally we'd want to
        # normalize these, but more importantly we don't want to break backwards
        # compatibility for either the SHA or MD5 users.
        md5_fp = fingerprint_from_ssh_pub_key(pub_key)
        sha256_fp = sha256_fingerprint_from_ssh_pub_key(pub_key)
        if (sha256_fp == fingerprint or
            md5_fp == fingerprint or
            "MD5:" + md5_fp == fingerprint):

            # if the user has given us sha256 fingerprint, canonicalize
            # it to the md5 fingerprint
            fingerprint = md5_fp
            break
    else:
        raise MantaError(
            "no '~/.ssh/*.pub' key found with fingerprint '%s'"
            % fingerprint)

    # XXX: pubkey should NOT be in PEM format.
    try:
        algo = ALGO_FROM_SSH_KEY_TYPE[pub_key.split()[0]]
    except KeyError:
        raise MantaError("Unsupported key type for: {}".format(key_id))

    priv_key_path = os.path.splitext(pub_key_path)[0]
    if not skip_priv_key:
        f = io.open(priv_key_path, 'rb')
        try:
            priv_key = f.read()
        finally:
            f.close()
    return dict(
        pub_key_path=pub_key_path,
        fingerprint=fingerprint,
        priv_key_path=priv_key_path,
        priv_key=priv_key,
        algorithm=algo)

def rsa_sig_from_agent_signed_response(response):
    msg = Message(response)
    algo = msg.get_string()
    signature = msg.get_string()

    return signature

def ecdsa_sig_from_agent_signed_response(response):
    msg = Message(response)
    algo = msg.get_text()
    sig = msg.get_binary()

    sig_msg = Message(sig)
    r = sig_msg.get_mpint()
    s = sig_msg.get_mpint()
    signature = encode_dss_signature(r, s)

    return signature

def ssh_key_info_from_key_data(key_id, priv_key=None):
    """Get/load SSH key info necessary for signing.

    @param key_id {str} Either a private ssh key fingerprint, e.g.
        'b3:f0:a1:6c:18:3b:42:63:fd:6e:57:42:74:17:d4:bc', or the path to
        an ssh private key file (like ssh's IdentityFile config option).
    @param priv_key {str} Optional. SSH private key file data (PEM format).
    @return {dict} with these keys:
        - type: "agent"
        - signer: Crypto signer class (a PKCS#1 v1.5 signer for RSA keys)
        - fingerprint: key md5 fingerprint
        - algorithm: See ALGO_FROM_SSH_KEY_TYPE for supported list.
        - ... some others added by `load_ssh_key()`
    """
    if FINGERPRINT_RE.match(key_id) and priv_key:
        key_info = {"fingerprint": key_id, "priv_key": priv_key}
    else:
        # Otherwise, we attempt to load necessary details from ~/.ssh.
        key_info = load_ssh_key(key_id)

    # Load a key signer.
    key = None
    try:
        key = serialization.load_pem_private_key(
            key_info["priv_key"],
            password=None,
            backend=default_backend())
    except TypeError as ex:
        log.debug("could not import key without passphrase (will "
            "try with passphrase): %s", ex)
        if "priv_key_path" in key_info:
            prompt = "Passphrase [%s]: " % key_info["priv_key_path"]
        else:
            prompt = "Passphrase: "
        for i in range(3):
            passphrase = getpass(prompt)
            if not passphrase:
                break
            try:
                key = serialization.load_pem_private_key(
                    key_info["priv_key"],
                    password=passphrase,
                    backend=default_backend())
            except ValueError:
                continue
            else:
                break
        if not key:
            details = ""
            if "priv_key_path" in key_info:
                details = " (%s)" % key_info["priv_key_path"]
            raise MantaError("could not import key" + details)

    # If load_ssh_key() wasn't run, set the algorithm here.
    if 'algorithm' not in key_info:
        if isinstance(key, ec.EllipticCurvePrivateKey):
            key_info['algorithm'] = ECDSA_ALGO_FROM_KEY_SIZE[str(key.key_size)]
        elif isinstance(key, rsa.RSAPrivateKey):
            key_info['algorithm'] = RSA_STR
        else:
            raise MantaError("Unsupported key type for: {}".format(key_id))

    key_info["signer"] = key
    key_info["type"] = "ssh_key"
    return key_info

def agent_key_info_from_key_id(key_id):
    """Find a matching key in the ssh-agent.

    @param key_id {str} Either a private ssh key fingerprint, e.g.
        'b3:f0:a1:6c:18:3b:42:63:fd:6e:57:42:74:17:d4:bc', or the path to
        an ssh private key file (like ssh's IdentityFile config option).
    @return {dict} with these keys:
        - type: "agent"
        - agent_key: paramiko AgentKey
        - fingerprint: key fingerprint
        - algorithm: "rsa-sha1"  Currently don't support DSA agent signing.
    """
    # Need the fingerprint of the key we're using for signing. If it
    # is a path to a priv key, then we need to load it.
    if not FINGERPRINT_RE.match(key_id):
        ssh_key = load_ssh_key(key_id, True)
        fingerprint = ssh_key["fingerprint"]
    else:
        fingerprint = key_id

    # Look for a matching fingerprint in the ssh-agent keys.
    keys = Agent().get_keys()

    for key in keys:
        raw_key = key.blob


        # The MD5 fingerprint functions return the hexdigest without the hash
        # algorithm prefix ("MD5:"), and the SHA256 functions return the
        # fingerprint with the prefix ("SHA256:").  Ideally we'd want to
        # normalize these, but more importantly we don't want to break backwards
        # compatibility for either the SHA or MD5 users.
        md5_fp = fingerprint_from_raw_ssh_pub_key(raw_key)
        sha_fp = sha256_fingerprint_from_raw_ssh_pub_key(raw_key)
        if (sha_fp == fingerprint or
            md5_fp == fingerprint or
            "MD5:" + md5_fp == fingerprint):

            # Canonicalize it to the md5 fingerprint.
            md5_fingerprint = md5_fp
            break
    else:
        raise MantaError('no ssh-agent key with fingerprint "%s"' %
                         fingerprint)

    return {
        "type": "agent",
        "agent_key": key,
        "fingerprint": md5_fingerprint,
        "algorithm": ALGO_FROM_SSH_KEY_TYPE[key.name]
    }

def ssh_key_sign(key_info, message):
    algo = key_info["algorithm"].split('-')
    hash_algo = algo[1]
    key_type = algo[0]

    hash_class = {
        "sha1": SHA1,
        "sha256": SHA256,
        "sha384": SHA384,
        "sha512": SHA512
    }[hash_algo]

    assert isinstance(message, bytes)

    if key_type == 'ecdsa':
        signed_raw = key_info["signer"].sign(
            message,
            ec.ECDSA(hash_class())
        )
    else:
        assert (key_type == 'rsa')
        signed_raw = key_info["signer"].sign(
            message,
            padding.PKCS1v15(),
            hash_class()
        )
    signed = base64.b64encode(signed_raw)
    return signed


#---- exports


class Signer(object):
    """A virtual base class for python-manta request signing."""

    def sign(self, s):
        """Sign the given string.

        @param s {str} The string to be signed.
        @returns (algorithm, key-md5-fingerprint, signature) {3-tuple}
            For example: `("rsa-sha256", "b3:f0:...:bc",
            "OXKzi5+h1aR9dVWHOu647x+ijhk...6w==")`.
        """
        raise NotImplementedError("this is a virtual base class")


class PrivateKeySigner(Signer):
    """Sign Manta requests with the given ssh private key.

    @param key_id {str} Either a private ssh key fingerprint, e.g.
        'b3:f0:a1:6c:18:3b:42:63:fd:6e:57:42:74:17:d4:bc', or the path to
        an ssh private key file (like ssh's IdentityFile config option).
    @param priv_key {str} Optional. SSH private key file data (PEM format).

    If a *fingerprint* is provided for `key_id` *and* `priv_key` is specified,
    then this is all the data required. Otherwise, this class will attempt
    to load required key data (both public and private key files) from
    keys in "~/.ssh/".
    """

    def __init__(self, key_id, priv_key=None):
        self.key_id = key_id
        self.priv_key = priv_key

    _key_info_cache = None

    def _get_key_info(self):
        """Get key info appropriate for signing."""
        if self._key_info_cache is None:
            self._key_info_cache = ssh_key_info_from_key_data(self.key_id,
                                                              self.priv_key)
        return self._key_info_cache

    def sign(self, s):
        if not isinstance(s, bytes):
            assert isinstance(s, str)
            s = s.encode("utf-8")

        key_info = self._get_key_info()

        assert key_info["type"] == "ssh_key"

        signed = ssh_key_sign(key_info, s)

        return (key_info["algorithm"], key_info["fingerprint"], signed)


class SSHAgentSigner(Signer):
    """Sign Manta requests using an ssh-agent.

    @param key_id {str} Either a private ssh key fingerprint, e.g.
        'b3:f0:a1:6c:18:3b:42:63:fd:6e:57:42:74:17:d4:bc', or the path to
        an ssh private key file (like ssh's IdentityFile config option).
    """

    def __init__(self, key_id):
        self.key_id = key_id

    _key_info_cache = None

    def _get_key_info(self):
        """Get key info appropriate for signing."""
        if self._key_info_cache is None:
            self._key_info_cache = agent_key_info_from_key_id(self.key_id)
        return self._key_info_cache

    def sign(self, s):

        if not isinstance(s, bytes):
            assert isinstance(s, str)
            s = s.encode("utf-8")

        key_info = self._get_key_info()
        assert key_info["type"] == "agent"

        response = key_info["agent_key"].sign_ssh_data(s)

        if re.search(r'^ecdsa-', key_info['algorithm']):
            signed_raw = ecdsa_sig_from_agent_signed_response(response)
        else:
            signed_raw = rsa_sig_from_agent_signed_response(response)

        signed = base64.b64encode(signed_raw)

        return (key_info["algorithm"], key_info["fingerprint"], signed)


class CLISigner(Signer):
    """Sign Manta requests using the SSH agent (if available and has the
    required key) or loading keys from "~/.ssh/*".
    """

    def __init__(self, key_id):
        self.key_id = key_id

    _key_info_cache = None

    def _get_key_info(self):
        """Get key info appropriate for signing: either from the ssh agent
        or from a private key.
        """
        if self._key_info_cache is not None:
            return self._key_info_cache

        errors = []

        # First try the agent.
        try:
            key_info = agent_key_info_from_key_id(self.key_id)
        except MantaError:
            _, ex, _ = sys.exc_info()
            errors.append(ex)
        else:
            self._key_info_cache = key_info
            return self._key_info_cache

        # Try loading from "~/.ssh/*".
        try:
            key_info = ssh_key_info_from_key_data(self.key_id)
        except MantaError:
            _, ex, _ = sys.exc_info()
            errors.append(ex)
        else:
            self._key_info_cache = key_info
            return self._key_info_cache

        raise MantaError("could not find key info for signing: %s" %
                         "; ".join(map(str, errors)))

    def sign(self, sigstr):
        if not isinstance(sigstr, bytes):
            assert isinstance(sigstr, str)
            sigstr = sigstr.encode("utf-8")

        key_info = self._get_key_info()
        log.debug("sign %r with %s key (algo %s, fp %s)", sigstr,
                  key_info["type"], key_info["algorithm"],
                  key_info["fingerprint"])

        if key_info["type"] == "agent":
            response = key_info["agent_key"].sign_ssh_data(sigstr)

            if re.search(r'^ecdsa-', key_info['algorithm']):
                signed_raw = ecdsa_sig_from_agent_signed_response(response)
            else:
                signed_raw = rsa_sig_from_agent_signed_response(response)

            signed = base64.b64encode(signed_raw)
        elif key_info["type"] == "ssh_key":
            signed = ssh_key_sign(key_info, sigstr)
        else:
            raise MantaError("internal error: unknown key type: %r" %
                             key_info["type"])

        return (key_info["algorithm"], key_info["fingerprint"], signed)
